pandas/tests/test_groupby.py

```diff
@@ -354,6 +354,41 @@ def test_nth_multi_index_as_expected(self):
                                          names=['A', 'B']))
         assert_frame_equal(result, expected)
 
+    def test_group_selection_cache(self):
+        # GH 12839 nth, head, and tail should return same result consistently
+        df = DataFrame([[1, 2], [1, 4], [5, 6]], columns=['A', 'B'])
+        expected = df.iloc[[0, 2]].set_index('A')
+
+        g = df.groupby('A')
+        g.head()
+        result = g.nth(0)
+        assert_frame_equal(result, expected)
+
+        g = df.groupby('A')
+        g.tail()
+        result = g.nth(0)
+        assert_frame_equal(result, expected)
+
+        g = df.groupby('A')
+        g.nth(0)
+        result = g.head(n=2)
+        assert_frame_equal(result, df)
+
+        g = df.groupby('A')
+        g.nth(0)
+        result = g.tail(n=2)
+        assert_frame_equal(result, df)
+
+        g = df.groupby('A')
+        g.head()
+        result = g.head(n=2)
+        assert_frame_equal(result, df)
+
```

Comment:
No, the issue was switching between two methods that either use a group selection (`nth`) or don't (`head` and `tail`). The bug is that once one of these methods are called, `_selected_obj` is placed into the cache either with or without the group selection (`self.obj[self._group_selection]` vs `self.obj`), and further retrievals from the cache will retrieve the same object without paying attention to whether the group selection has changed or not. The proposed pull request clears the cache for `_selected_obj` whenever the group selection changes. This happens in `_set_selection_from_grouper` in `nth`, and it also needs to happen at the beginning of `head` and `tail`.